### PR TITLE
BugFix: fatal error when the delta writer is busy

### LINKED ISSUE #3877: stream load leads BE fatal error
The `bthread::ExecutionQueue` used in StarRocks may get fatal error when the number of tasks in the queue is overloading.

### Steps to reproduce the behavior (Required)
1. creating table with massive buckets
To make it easier to reproduce, I create a table with 2000 buckets.
3. making stream load to the table created before
To make it easier to reproduce, I make stream load with 32 parallelism and 10MB batch.

### Expected behavior (Required)
The BE runs properly.
### Real behavior (Required)
The BE get fatal error.
```
W0305 14:53:20.082063 19490 async_delta_writer_executor.h:32] Service unavailable: Thread pool is at capacity (16/16 tasks running, 40960/40960 tasks queued)
F0305 14:53:20.082093 19490 execution_queue.cpp:120] Fail to submit task: Resource temporarily unavailable [11]
```
```
*** Check failure stack trace: ***
    @          0x3b2f86d  google::LogMessage::Fail()
    @          0x3b31b49  google::LogMessage::SendToLog()
    @          0x3b2f3e9  google::LogMessage::Flush()
    @          0x3b2f5f2  google::LogMessage::~LogMessage()
    @          0x3c0c01c  bthread::ExecutionQueueBase::start_execute()
    @          0x30cf32f  starrocks::vectorized::AsyncDeltaWriter::write()
    @          0x1e2acec  starrocks::TabletsChannel::add_chunk()
    @          0x1e27565  starrocks::LoadChannel::add_chunk()
    @          0x1e2405b  starrocks::LoadChannelMgr::add_chunk()
    @          0x3c5f4fe  brpc::policy::ProcessRpcRequest()
    @          0x3c55f07  brpc::ProcessInputMessage()
    @          0x3c56db3  brpc::InputMessenger::OnNewMessages()
    @          0x3cfde2e  brpc::Socket::ProcessEvent()
    @          0x3da2c6f  bthread::TaskGroup::task_runner()
    @          0x3d97191  bthread_make_fcontext
```
### StarRocks version (Required)

main, branch-2.1

### PR DESCRIPTION
**Pull request by rickif**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3877

## Problem Summary(Required) ：

This PR make retry when the `AsyncDeltaWriter` is busy to avoid the `bthread`'s fatal error.

### CODE REVIEW

File: be/src/storage/async_delta_writer_executor.h [hunk 1]
```diff
@@ -7,28 +7,36 @@ DIAGNOSTIC_IGNORE("-Wclass-memaccess")
...
     }
 
     int submit(void* (*fn)(void*), void* args) override {
-        auto st = _thread_pool->submit_func([=]() { fn(args); });
+        Status st;
+        while (true) {
+            st = _thread_pool->submit_func([=]() { fn(args); });
+            if (!st.is_service_unavailable()) break;
+            LOG(INFO) << "async_delta_writer is busy, retry after ms: " << RETRY_INTERVAL_MS;
+            SleepFor(MonoDelta::FromMilliseconds(RETRY_INTERVAL_MS));
```
Comment (sduzh) [inline be/src/storage/async_delta_writer_executor.h:38]: According to my knowledge, this method may be involked in a bthread, calling `SleepFor` in bthread is bad. Maybe you can return EAGAIN on service unavailable and let the caller do the retry.

File: be/src/storage/async_delta_writer_executor.h [hunk 2]
```diff
@@ -7,28 +7,36 @@ DIAGNOSTIC_IGNORE("-Wclass-memaccess")
 DIAGNOSTIC_POP
 
 #include "common/config.h"
+#include "util/monotime.h"
 #include "util/threadpool.h"
 
 namespace starrocks {
+const int64_t RETRY_INTERVAL_MS = 50;
```
Comment (sduzh) [inline be/src/storage/async_delta_writer_executor.h:14]:
```suggestion
const int64_t kRetryIntervalMs = 50;
```

File: be/src/storage/async_delta_writer_executor.h [hunk 3]
```diff
@@ -7,28 +7,36 @@ DIAGNOSTIC_IGNORE("-Wclass-memaccess")
...
                 .build(&_thread_pool);
     }
 
     int submit(void* (*fn)(void*), void* args) override {
-        auto st = _thread_pool->submit_func([=]() { fn(args); });
+        Status st;
+        while (true) {
+            st = _thread_pool->submit_func([=]() { fn(args); });
+            if (!st.is_service_unavailable()) break;
+            LOG(INFO) << "async_delta_writer is busy, retry after ms: " << RETRY_INTERVAL_MS;
```
Comment (sduzh) [inline be/src/storage/async_delta_writer_executor.h:37]:
```suggestion
            LOG(INFO) << "async_delta_writer is busy, retry after " << RETRY_INTERVAL_MS << "ms";
```

File: be/src/storage/async_delta_writer_executor.h (hunk 1 above)
Comment (rickif) [reply be/src/storage/async_delta_writer_executor.h:38]: `ExecutionQueueBase` calls `bthread::Executor::submit` and `PLOG(FATAL)` if  get any error.
https://github.com/apache/incubator-brpc/blob/1ad1cc57ecc50d4dd631e32781459830f75ca23a/src/bthread/execution_queue.cpp#L120

Comment (sduzh) [reply be/src/storage/async_delta_writer_executor.h]: If so, use [bthread_usleep](https://github.com/apache/incubator-brpc/blob/37a21f44f97ffe8f17bd542c2de4f974ddd16bbe/src/bthread/bthread.h#L156)

Comment (rickif) [reply be/src/storage/async_delta_writer_executor.h]: I agree and have updated the code.

**Review by sduzh — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

File: be/src/storage/async_delta_writer_executor.h (hunk 1 above)
Comment (chaoyli) [reply be/src/storage/async_delta_writer_executor.h]: If the brpc can not succeed, will it be an endless loop?
Does it seem like a BRPC bug?

Comment (sduzh) [reply be/src/storage/async_delta_writer_executor.h]: The problem is `_thread_pool` may return an error status on busy, it has nothing to do with brpc.

Comment (rickif) [reply be/src/storage/async_delta_writer_executor.h]: > `ExecutionQueueBase` calls `bthread::Executor::submit` and `PLOG(FATAL)` if get any error. https://github.com/apache/incubator-brpc/blob/1ad1cc57ecc50d4dd631e32781459830f75ca23a/src/bthread/execution_queue.cpp#L120

The direct reason is that the user-defined `bthread::Executor::submit` cannot return non-zero. I think it's not good but seems no better solution.

Comment (sduzh) [reply be/src/storage/async_delta_writer_executor.h]: Got it. you can try to connect the brpc community ask if it's to change the log level from FATAL to ERROR.

**Comment by rickif** [comment]
@mergifyio backport branch-2.1

**Pull request merged**